src/stream/src/executor/aggregation/agg_impl/approx_distinct_utils.rs

```diff
@@ -0,0 +1,203 @@
+// Copyright 2022 Singularity Data
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::hash_map::DefaultHasher;
+use std::hash::{Hash, Hasher};
+
+use dyn_clone::DynClone;
+use itertools::Itertools;
+use risingwave_common::array::stream_chunk::Ops;
+use risingwave_common::array::*;
+use risingwave_common::buffer::Bitmap;
+use risingwave_common::types::{Datum, DatumRef, Scalar, ScalarImpl};
+
+use crate::executor::aggregation::agg_impl::StreamingAggImpl;
+use crate::executor::StreamExecutorResult;
+
+const INDEX_BITS: u8 = 16; // number of bits used for finding the index of each 64-bit hash
+const NUM_OF_REGISTERS: u32 = 1 << INDEX_BITS; // number of registers available
+const COUNT_BITS: u8 = 64 - INDEX_BITS; // number of non-index bits in each 64-bit hash
+
+// Approximation for bias correction for 16384 registers. See "HyperLogLog: the analysis of a
+// near-optimal cardinality estimation algorithm" by Philippe Flajolet et al.
+const BIAS_CORRECTION: f64 = 0.7213 / (1. + (1.079 / NUM_OF_REGISTERS as f64));
+
+pub(super) trait RegisterBucket {
+    fn new() -> Self;
+
+    /// Increments or decrements the bucket at `index` depending on the state of `is_insert`.
+    /// Returns an Error if `index` is invalid or if inserting will cause an overflow in the bucket.
+    fn update_bucket(&mut self, index: usize, is_insert: bool) -> StreamExecutorResult<()>;
+
+    /// Gets the number of the maximum bucket which has a count greater than zero.
+    fn get_max(&self) -> StreamExecutorResult<u8>;
+}
+
+/// `StreamingApproxDistinct` approximates the count of non-null rows using a modified version
+/// of the `HyperLogLog` algorithm. Each `RegisterBucket` stores a count of how many hash values
+/// have x trailing zeroes for all x from 1-64. This allows the algorithm to support insertion and
+/// deletion, but uses up more memory and limits the number of rows that can be counted.
+///
+/// `StreamingApproxDistinct` can count up to a total of 2^64 unduplicated rows.
+///
+/// The estimation error for `HyperLogLog` is 1.04/sqrt(num of registers). With 2^16 registers this
+/// is ~1/256, or about 0.4%. The memory usage for the default choice of parameters is about
+/// (1024 + 24) bits * 2^16 buckets, which is about 8.58 MB.
+pub(super) trait StreamingApproxDistinct: Sized {
```

Comment:
Overall naming convention is a bit weird.

Can we stick with either `StreamingApproxDistinct` or `StreamingApproxCountDistinct` as trait, and then the impls become `Updateable_` and `AppendOnly_`?

Reply:
I'd go with `StreamingApproxCountDistinct` for trait.

Reply:
> Overall naming convention is a bit weird.
> 
> Can we stick with either `StreamingApproxDistinct` or `StreamingApproxCountDistinct` as trait, and then the impls become `Updateable_` and `AppendOnly_`?

fixed

Reply:
> I'd go with `StreamingApproxCountDistinct` for trait.

I use `StreamingApproxDistinct` because there are not other things whose distinct count is possible to be estimated. We might approx `ApproxMedian` but never `ApproxMedianDistinct`.

Reply:
IMO naming is not just about discriminating from other names, but also describing the item itself. `StreamingApproxCountDistinct` might be more descriptive.

Reply:
> IMO naming is not just about discriminating from other names, but also describing the item itself. `StreamingApproxCountDistinct` might be more descriptive.

fixed